content/knowledge_base/flow/cadence/reference.md

```diff
@@ -0,0 +1,330 @@
+factorial(-2)
+```
+
+## Switches
+
+Different from the main-stream languages, switches in cadence do not implicit fallthrough.
+
+## Composite Types
+
+Composite types can only be declared within a contract and nowhere else.
+
+There are two types:
+
+* Structures
+    * They are copied.
+    * They are value types.
+* Resources
+    * They are moved.
+    * They are linear types.
+
+## Equatable Interface
+
+```rust
+struct interface Equatable {
+    pub fun equals(_ other: {Equatable}): Bool
+}
+```
+
+## Hashable Interface
+
+```rust
+struct interface Hashable: Equatable {
+    pub hashValue: Int
+}
+```
+
+## Restricted Types
+
+Resources and structs can be used with the restricted types.
+For example, a Resource that implements the interface `Balance` can be accessed by accessing only the functions and/or fields available in Balance.
+
+```rust
+resource interface HasCount {
+    pub let count: Int
+}
+
+pub resource Counter: HasCount {
+    pub var count: Int
+
+    init(count: Int) {
+        self.count = count
+    }
+
+    pub fun increment() {
+        self.count = self.count + 1
+    }
+}
+
+let counter: @Counter <- create Counter(count: 42)
+
+counter.count  // is `42`
+
+counter.increment()
+
+counter.count  // is `43`
+
+// Move the resource in variable `counter` to a new variable `restrictedCounter`,
+// but typed with the restricted type `Counter{HasCount}`:
+// The variable may hold any `Counter`, but only the functionality
+// defined in the given restriction, the interface `HasCount`, may be accessed
+//
+let restrictedCounter: @Counter{HasCount} <- counter
+
+// Invalid: Only functionality of restriction `Count` is available,
+// i.e. the read-only field `count`, but not the function `increment` of `Counter`
+//
+restrictedCounter.increment()
+```
+
+## Events
+
+Events can only be declared within contracts and cannot use resources as parameters or outputs.
+
+```rust
+// Invalid: An event cannot be declared globally
+//
+event GlobalEvent(field: Int)
+
+pub contract Events {
+    // Event with explicit argument labels
+    //
+    event BarEvent(labelA fieldA: Int, labelB fieldB: Int)
+
+    // Invalid: A resource type is not allowed to be used
+    // because it would be moved and lost
+    //
+    event ResourceEvent(resourceField: @Vault)
+}
+```
+
+### Emitting events
+
+To Emit an event the keyword `emit` is used. For example, emitting an event of the type `Test(field: Int)`: `emit Test(1)`.
+
+### Core Events
+
+There are some core events built-in in cadence.
+
+* `pub event AccountCreated(address: Address)`
+* `pub event AccountKeyAdded(address: Address, publicKey: [UInt8])`
+* `pub event AccountKeyRemoved(address: Address, publicKey: [UInt8])`
+* `pub event AccountContractAdded(address: Address, codeHash: [UInt8], contract: String)`
+* `pub event AccountContractUpdated(address: Address, codeHash: [UInt8], contract: String)`
+* `pub event AccountContractRemoved(address: Address, codeHash: [UInt8], contract: String)`
```

Comment:
I would suggest linking each of those bullet points to the related event's code since IIRC it is in the reference or at least could be in the code.